lib/src/model/tournament/tournament_controller.dart

```diff
@@ -0,0 +1,150 @@
+import 'dart:async';
+
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:lichess_mobile/src/model/common/id.dart';
+import 'package:lichess_mobile/src/model/common/socket.dart';
+import 'package:lichess_mobile/src/model/tournament/tournament.dart';
+import 'package:lichess_mobile/src/model/tournament/tournament_repository.dart';
+import 'package:lichess_mobile/src/network/socket.dart';
+import 'package:logging/logging.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+part 'tournament_controller.freezed.dart';
+part 'tournament_controller.g.dart';
+
+final _logger = Logger('TournamentController');
+
+@riverpod
+class TournamentController extends _$TournamentController {
+  StreamSubscription<SocketEvent>? _socketSubscription;
+
+  late final SocketClient _socketClient;
+
+  // If we join/leave to often, the server blocks us from joining, see [TournamentMe.pauseDelay].
+  // However, there's no "reload" event from the socket once we're able to join again,
+  // so we manually have to set this timer to schedule a reload once we can join again.
+  Timer? _pauseDelayTimer;
+
+  @override
+  Future<TournamentState> build(TournamentId id) async {
+    final state = await _loadTournament(id, standingsPage: 1);
+
+    final socketPool = ref.watch(socketPoolProvider);
+    _socketClient = socketPool.open(Uri(path: '/tournament/$id/socket/v6'));
+    ref.onDispose(() {
```

Comment:
Better to put the `ref.onDispose` at the top of `build` (not after any `await`) to ensure the callback is called.